libkbfs/block_tree.go

```diff
@@ -794,3 +799,114 @@ func (bt *blockTree) markParentsDirty(parentBlocks []parentBlockAndChildIndex) (
 	}
 	return dirtyPtrs, unrefs, nil
 }
+
+type makeSyncFunc func(ptr BlockPointer) func() error
+
+// readyHelper takes a set of paths from a root down to a child block,
+// and readies all the blocks represented in those paths.  If the
+// caller wants leaf blocks readied, then the last element of each
+// slice in `pathsFromRoot` should contain a leaf block, with a child
+// index of -1.  It's assumed that all slices in `pathsFromRoot` have
+// the same size. This function returns a map pointing from the new
+// block info from any readied block to its corresponding old block
+// pointer.
+func (bt *blockTree) readyHelper(
+	ctx context.Context, id tlf.ID, bcache BlockCache, bops BlockOps,
+	bps *blockPutState, pathsFromRoot [][]parentBlockAndChildIndex,
+	makeSync makeSyncFunc) (map[BlockInfo]BlockPointer, error) {
+	oldPtrs := make(map[BlockInfo]BlockPointer)
+	newPtrs := make(map[BlockPointer]bool)
+
+	// Starting from the leaf level, ready each block at each level,
+	// and put the new BlockInfo into the parent block at the level
+	// above.  At each level, only ready each block once. Don't ready
+	// the root block though; the folderUpdatePrepper code will do
+	// that.
+	for level := len(pathsFromRoot[0]) - 1; level > 0; level-- {
+		for i := 0; i < len(pathsFromRoot); i++ {
+			// Ready the dirty block.
+			pb := pathsFromRoot[i][level]
```

Comment:
I know this is just a refactor, but are we guaranteed that each `pathsFromRoot[i]` will have `len(pathsFromRoot[0])` levels?

Reply:
That's a great question.  I think the answer is yes, because of the way making new blocks work.  If a block is split, and all parent blocks are full, we add a new block level above the top of the whole tree, and then create a new branch on the right side extending all the way down.  So the tree should have a balanced number of levels.

So actually I think comments I've written in the past, about the directory tree not necessarily being balanced, weren't correct.